edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs

```diff
@@ -0,0 +1,938 @@
+                            break;
+
+                        case WatchEventType.Deleted:
+                            {
+                                // Delete the deployment.
+                                // Delete any services.
+                                IEnumerable<Task<V1Status>> removeServiceTasks = currentServices.Items.Select(i => this.client.DeleteNamespacedServiceAsync(i.Metadata.Name, this.k8sNamespace, new V1DeleteOptions()));
+                                await Task.WhenAll(removeServiceTasks);
+                                IEnumerable<Task<V1Status>> removeDeploymentTasks = currentDeployments.Items.Select(
+                                    d => this.client.DeleteNamespacedDeployment1Async(
+                                        d.Metadata.Name,
+                                        this.k8sNamespace,
+                                        new V1DeleteOptions(propagationPolicy: "Foreground"),
+                                        propagationPolicy: "Foreground"));
+                                await Task.WhenAll(removeDeploymentTasks);
+                                this.currentModules = ModuleSet.Empty;
+                            }
+
+                            break;
+                        case WatchEventType.Error:
+                            Events.DeploymentError();
+                            break;
+                    }
+                }
+            }
+            else
+            {
+                Events.DeploymentNameMismatch(customObject.Metadata.Name, this.resourceName);
+            }
+        }
+
+        private string DeploymentName(string moduleId) => KubeUtils.SanitizeK8sValue(moduleId);
+
+        private async void ManageDeployments(V1ServiceList currentServices, V1DeploymentList currentDeployments, EdgeDeploymentDefinition<TConfig> customObject)
+        {
+            var desiredModules = ModuleSet.Create(customObject.Spec.ToArray());
+            var moduleIdentities = await this.moduleIdentityLifecycleManager.GetModuleIdentitiesAsync(desiredModules, this.currentModules);
+
+            // Pull current configuration from annotations.
+            Dictionary<string, string> currentV1ServicesFromAnnotations = this.GetCurrentServiceConfig(currentServices);
+            // strip out edgeAgent so edgeAgent doesn't update itself.
+            // TODO: remove this filter.
+            var agentDeploymentName = this.DeploymentName(CoreConstants.EdgeAgentModuleName);
+            Dictionary<string, string> currentDeploymentsFromAnnotations = this.GetCurrentDeploymentConfig(currentDeployments)
+                .ToDictionary(
+                    pair => pair.Key,
+                    pair => pair.Value);
+
+            var desiredServices = new List<V1Service>();
+            var desiredDeployments = new List<V1Deployment>();
+            foreach (KubernetesModule<TConfig> module in customObject.Spec)
```

Comment:
I think this method needs to be split into smaller methods so it's easier to follow and to test

Reply:
I agree @ancaantochi ! I am saving that for when we do the larger async/await refactor in the crd watch.